File: Windows/Type4Me/Session/SoundFeedback.cs
using System.IO;
using System.Media;
using Type4Me.Audio;
using Type4Me.Services;

namespace Type4Me.Session;

/// <summary>
/// Synthesized audio feedback tones.
/// Uses programmatic WAV generation — no external sound files needed.
/// </summary>
public static class SoundFeedback
{
    private record ToneSpec(
        (double Frequency, double Duration)[] Tones,
        float Volume,
        string Label);

    private static bool _hasWarmedUp;
    private static readonly Dictionary<string, byte[]> _cachedWavs = new();

    private static readonly ToneSpec StartSpec = new(
        [(587, 0.06), (880, 0.09)],
        0.52f, "start");

    private static readonly ToneSpec StopSpec = new(
        [(740, 0.04), (1175, 0.06)],
        0.3f, "stop");

    private static readonly ToneSpec ErrorSpec = new(
        [(330, 0.08), (220, 0.1)],
        0.35f, "error");

    // ── Public API ─────────────────────────────────────────

    public static void WarmUp()
    {
        if (_hasWarmedUp) return;
        _hasWarmedUp = true;
        DebugFileLogger.Log("[SoundFeedback] warmUp");

        // Pre-generate all WAV data
        GetOrBuildWav(StartSpec);
        GetOrBuildWav(StopSpec);
        GetOrBuildWav(ErrorSpec);
    }

    public static void PlayStart()
    {
        DebugFileLogger.Log("[SoundFeedback] playStart");
        Play(StartSpec);
    }

    public static void PlayStop()
    {
        DebugFileLogger.Log("[SoundFeedback] playStop");
        Play(StopSpec);
    }

    public static void PlayError()
    {
        DebugFileLogger.Log("[SoundFeedback] playError");
        Play(ErrorSpec);
    }

    // ── Playback ───────────────────────────────────────────

    private static void Play(ToneSpec spec)
    {
        try
        {
            var wav = GetOrBuildWav(spec);
            using var ms = new MemoryStream(wav);
            using var player = new SoundPlayer(ms);
            player.Play();
        }
        catch (Exception ex)
        {
            DebugFileLogger.Log($"[SoundFeedback] {spec.Label} play failed: {ex.Message}");
            SystemSounds.Beep.Play();
        }
    }

    // ── Tone Synthesis ─────────────────────────────────────

    private static byte[] GetOrBuildWav(ToneSpec spec)
    {
        if (_cachedWavs.TryGetValue(spec.Label, out var cached))
            return cached;

        var wav = BuildToneData(spec);
        _cachedWavs[spec.Label] = wav;
        return wav;
    }

    private static byte[] BuildToneData(ToneSpec spec)
    {
        const int sampleRate = 44100;
        var samples = new List<short>();

        foreach (var (frequency, duration) in spec.Tones)
        {
            int frameCount = (int)(duration * sampleRate);
            for (int i = 0; i < frameCount; i++)
            {
                double t = (double)i / sampleRate;
                double envelope = Math.Sin(Math.PI * t / duration);
                double value = Math.Sin(2.0 * Math.PI * frequency * t) * envelope * 0.5 * spec.Volume;
                samples.Add((short)(value * 32767));
            }
        }

        return WavEncoder.EncodeFromSamples(samples.ToArray(), sampleRate);
    }
}

File: MacOS/Type4Me/Session/SoundFeedback.swift
import AppKit
import AVFoundation

/// Synthesized audio feedback tones.
/// Uses programmatic WAV generation — no external sound files needed.
enum SoundFeedback {

    private enum Delivery {
        case afplay
        case cachedPlayer
    }

    private struct ToneSpec {
        let tones: [(frequency: Double, duration: Double)]
        let volume: Float
        let label: String
        let delivery: Delivery
    }

    nonisolated(unsafe) private static var hasWarmedUp = false
    nonisolated(unsafe) private static var cachedPlayers: [String: AVAudioPlayer] = [:]
    nonisolated(unsafe) private static var activeProcesses: [Process] = []

    private static let startSpec = ToneSpec(
        tones: [
            (frequency: 587, duration: 0.06),
            (frequency: 880, duration: 0.09),
        ],
        volume: 0.52,
        label: "start",
        delivery: .afplay
    )

    private static let stopSpec = ToneSpec(
        tones: [
            (frequency: 740, duration: 0.04),
            (frequency: 1175, duration: 0.06),
        ],
        volume: 0.3,
        label: "stop",
        delivery: .cachedPlayer
    )

    private static let errorSpec = ToneSpec(
        tones: [
            (frequency: 330, duration: 0.08),
            (frequency: 220, duration: 0.1),
        ],
        volume: 0.35,
        label: "error",
        delivery: .cachedPlayer
    )

    // MARK: - Public API

    static func warmUp() {
        DispatchQueue.main.async {
            guard !hasWarmedUp else { return }
            hasWarmedUp = true
            NSLog("[SoundFeedback] warmUp")
            DebugFileLogger.log("sound warmUp")
            _ = try? soundFileURL(for: startSpec)
            preparePlayersIfNeeded()
        }
    }

    /// Softer ascending chime for recording start.
    static func playStart() {
        NSLog("[SoundFeedback] playStart")
        DebugFileLogger.log("sound playStart invoked")
        play(spec: startSpec, retryCount: 2)
    }

    /// Crisper, more decisive double tone for recording stop.
    static func playStop() {
        NSLog("[SoundFeedback] playStop")
        DebugFileLogger.log("sound playStop invoked")
        play(spec: stopSpec)
    }

    /// Low descending fifth (330→220 Hz). Unmistakable but not harsh.
    static func playError() {
        NSLog("[SoundFeedback] playError")
        DebugFileLogger.log("sound playError invoked")
        play(spec: errorSpec)
    }

    // MARK: - Synthesis

    private static func play(
        spec: ToneSpec,
        retryCount: Int = 0,
        forcedVolume: Float? = nil
    ) {
        DispatchQueue.main.async {
            do {
                let didPlay: Bool
                switch spec.delivery {
                case .afplay:
                    didPlay = try playWithAFPlay(spec: spec)
                case .cachedPlayer:
                    didPlay = try playWithCachedPlayer(spec: spec, forcedVolume: forcedVolume)
                }

                NSLog("[SoundFeedback] %@ play() => %@", spec.label, didPlay ? "true" : "false")
                DebugFileLogger.log("sound \(spec.label) play() => \(didPlay)")
                guard didPlay else {
                    if retryCount > 0 {
                        NSLog("[SoundFeedback] %@ retry scheduled (%d left)", spec.label, retryCount)
                        DebugFileLogger.log("sound \(spec.label) retry scheduled, remaining=\(retryCount)")
                        DispatchQueue.main.asyncAfter(deadline: .now() + 0.12) {
                            play(spec: spec, retryCount: retryCount - 1, forcedVolume: forcedVolume)
                        }
                    } else {
                        NSLog("[SoundFeedback] %@ falling back to NSSound.beep()", spec.label)
                        DebugFileLogger.log("sound \(spec.label) fallback to NSSound.beep()")
                        NSSound.beep()
                    }
                    return
                }
            } catch {
                NSLog("[SoundFeedback] %@ init failed: %@", spec.label, String(describing: error))
                DebugFileLogger.log("sound \(spec.label) init failed: \(String(describing: error))")
                NSSound.beep()
            }
        }
    }

    private static func preparePlayersIfNeeded() {
        do {
            _ = try preparedPlayer(for: stopSpec)
            _ = try preparedPlayer(for: errorSpec)
        } catch {
            DebugFileLogger.log("sound preparePlayersIfNeeded failed: \(String(describing: error))")
        }
    }

    private static func preparedPlayer(for spec: ToneSpec) throws -> AVAudioPlayer {
        if let player = cachedPlayers[spec.label] {
            return player
        }

        let player = try AVAudioPlayer(data: buildToneData(for: spec))
        player.prepareToPlay()
        cachedPlayers[spec.label] = player
        return player
    }

    private static func playWithCachedPlayer(spec: ToneSpec, forcedVolume: Float?) throws -> Bool {
        let player = try preparedPlayer(for: spec)
        player.stop()
        player.currentTime = 0
        player.numberOfLoops = 0
        player.volume = forcedVolume ?? spec.volume
        player.prepareToPlay()
        return player.play()
    }

    private static func playWithAFPlay(spec: ToneSpec) throws -> Bool {
        let process = Process()
        process.executableURL = URL(fileURLWithPath: "/usr/bin/afplay")
        process.arguments = [try soundFileURL(for: spec).path]
        process.terminationHandler = { finished in
            DispatchQueue.main.async {
                activeProcesses.removeAll { $0 === finished }
                DebugFileLogger.log("sound \(spec.label) afplay terminated status=\(finished.terminationStatus)")
            }
        }
        try process.run()
        activeProcesses.append(process)
        DebugFileLogger.log("sound \(spec.label) afplay launched pid=\(process.processIdentifier)")
        return process.isRunning
    }

    private static func soundFileURL(for spec: ToneSpec) throws -> URL {
        let dir = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
            .appendingPathComponent("Type4Me", isDirectory: true)
        try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        let url = dir.appendingPathComponent("sound-\(spec.label).wav")
        if !FileManager.default.fileExists(atPath: url.path) {
            try buildToneData(for: spec).write(to: url, options: .atomic)
        }
        return url
    }

    private static func buildToneData(for spec: ToneSpec) -> Data {
        let sampleRate = 44100.0
        var samples = [Int16]()

        for tone in spec.tones {
            let frameCount = Int(tone.duration * sampleRate)
            for i in 0..<frameCount {
                let t = Double(i) / sampleRate
                let envelope = sin(.pi * t / tone.duration)
                let value = sin(2.0 * .pi * tone.frequency * t) * envelope * 0.5
                samples.append(Int16(value * 32767))
            }
        }

        return buildWAV(samples: samples, sampleRate: Int(sampleRate))
    }

    // MARK: - WAV Builder

    private static func buildWAV(samples: [Int16], sampleRate: Int) -> Data {
        let dataSize = samples.count * 2
        let fileSize = 36 + dataSize

        var wav = Data(capacity: 44 + dataSize)

        // RIFF header
        wav.append(contentsOf: [0x52, 0x49, 0x46, 0x46]) // "RIFF"
        wav.appendLE(UInt32(fileSize))
        wav.append(contentsOf: [0x57, 0x41, 0x56, 0x45]) // "WAVE"

        // fmt chunk (16 bytes, PCM)
        wav.append(contentsOf: [0x66, 0x6D, 0x74, 0x20]) // "fmt "
        wav.appendLE(UInt32(16))
        wav.appendLE(UInt16(1))                    // PCM format
        wav.appendLE(UInt16(1))                    // mono
        wav.appendLE(UInt32(sampleRate))
        wav.appendLE(UInt32(sampleRate * 2))       // byte rate
        wav.appendLE(UInt16(2))                    // block align
        wav.appendLE(UInt16(16))                   // bits per sample

        // data chunk
        wav.append(contentsOf: [0x64, 0x61, 0x74, 0x61]) // "data"
        wav.appendLE(UInt32(dataSize))
        for sample in samples {
            wav.appendLE(sample)
        }

        return wav
    }
}

// MARK: - Data Helper

private extension Data {
    mutating func appendLE<T: FixedWidthInteger>(_ value: T) {
        Swift.withUnsafeBytes(of: value.littleEndian) { append(contentsOf: $0) }
    }
}
